File: apps/mobile/src/lib/crashReporting.ts
/**
 * Crash reporting (Sentry) — on only when `EXPO_PUBLIC_SENTRY_DSN` is set at
 * build time, and never with personal data.
 *
 * The server's rule (02 §8, `app/observability/crash_reporting.py`) holds here
 * too: a crash report is a log line that leaves the phone, so it carries no
 * emails, no tokens, no URL query strings (an upload URL's query IS its
 * signature) and no request bodies. Sentry's own switches do most of it;
 * `scrubEvent` is the part that does not depend on remembering which switch
 * covers which field, and it returns a NEW event — what it is handed is not
 * touched.
 *
 * Off by default, like the API's: a development build, Expo Go and Jest send
 * nothing, and `wrapRoot` hands the layout back unwrapped.
 *
 * NOTE: `process.env.EXPO_PUBLIC_*` is inlined at BUILD time by
 * babel-preset-expo (see `src/lib/api.ts`). Setting the DSN after Metro has
 * started does nothing; rebuild. The default parameters below are written as
 * the literal member expressions for exactly that reason.
 */
import * as Sentry from '@sentry/react-native';
import type { ComponentType } from 'react';

const FILTERED = '[Filtered]';

/** Keys whose VALUE is a secret, wherever they appear. */
const SECRET_KEY = /password|passwd|secret|token|authorization|cookie|api_?key|signature/i;

/** Dropped outright: a request's body and cookies, a breadcrumb's query. */
const DROPPED_KEYS = new Set(['data', 'cookies', 'query_string', 'http.query', 'http.fragment']);

const EMAIL = /[\w.+-]+@[\w-]+(?:\.[\w-]+)+/g;
const JWT = /eyJ[\w-]+\.[\w-]+\.[\w-]+/g;
const URL_QUERY = /(https?:\/\/[^\s?#"']+)[?#][^\s"']*/g;

function scrubText(text: string): string {
  return text.replace(URL_QUERY, '$1').replace(JWT, '[token]').replace(EMAIL, '[email]');
}

function scrubValue(value: unknown, dropData: boolean): unknown {
  if (typeof value === 'string') return scrubText(value);
  if (Array.isArray(value)) return value.map((v) => scrubValue(v, dropData));
  if (value && typeof value === 'object') {
    const out: Record<string, unknown> = {};
    for (const [key, v] of Object.entries(value as Record<string, unknown>)) {
      if (dropData && DROPPED_KEYS.has(key)) continue;
      out[key] = SECRET_KEY.test(key) ? FILTERED : scrubValue(v, false);
    }
    return out;
  }
  return value;
}

/** Recursively: secrets filtered, emails, tokens and URL queries replaced. */
function scrub<T>(value: T): T {
  return scrubValue(value, false) as T;
}

type AnyEvent = Record<string, unknown> & {
  request?: Record<string, unknown>;
  user?: Record<string, unknown>;
};

/** `beforeSend`: the event without personal data. Never drops it. */
export function scrubEvent<E extends AnyEvent>(event: E): E {
  const cleaned = scrub(event) as AnyEvent;
  const out: AnyEvent = { ...cleaned };
  if (cleaned.request) {
    const { headers: _headers, ...rest } = scrubValue(cleaned.request, true) as Record<string, unknown>;
    // No headers at all: the client's own are the bearer token and nothing
    // a crash needs.
    out.request = rest;
  }
  if (cleaned.user) {
    // The id is how a report meets a support request; nothing else.
    out.user = 'id' in cleaned.user ? { id: cleaned.user.id } : {};
  }
  return out as E;
}

/** `beforeBreadcrumb`: the same rules, at the moment a breadcrumb is recorded. */
export function scrubBreadcrumb<B extends Record<string, unknown>>(breadcrumb: B): B {
  const cleaned = scrub(breadcrumb) as Record<string, unknown>;
  const data = cleaned.data;
  if (data && typeof data === 'object') {
    return { ...cleaned, data: scrubValue(data, true) } as unknown as B;
  }
  return cleaned as B;
}

let enabled = false;

/** Whether `initCrashReporting` turned Sentry on in this process. */
export const crashReportingEnabled = () => enabled;

export function initCrashReporting(
  dsn: string | undefined = process.env.EXPO_PUBLIC_SENTRY_DSN,
  environment: string | undefined = process.env.EXPO_PUBLIC_SENTRY_ENVIRONMENT,
): boolean {
  if (!dsn) return false;
  Sentry.init({
    dsn,
    environment: environment || (__DEV__ ? 'development' : 'production'),
    sendDefaultPii: false,
    // Pictures of the screen are pictures of somebody's meals and body.
    attachScreenshot: false,
    attachViewHierarchy: false,
    // Would attach request and response details of every failed call.
    enableCaptureFailedRequests: false,
    // Errors only: tracing is a second data stream with its own PII surface.
    tracesSampleRate: 0,
    beforeSend: (event) => scrubEvent(event as unknown as AnyEvent) as unknown as typeof event,
    beforeBreadcrumb: (breadcrumb) =>
      scrubBreadcrumb(breadcrumb as unknown as Record<string, unknown>) as unknown as typeof breadcrumb,
  });
  enabled = true;
  return true;
}

/**
 * The root layout, wrapped when crash reporting is on (touch breadcrumbs and
 * the root boundary). Unwrapped otherwise, so development and Jest render the
 * same tree they always did.
 */
export function wrapRoot<P extends Record<string, unknown>>(
  Root: ComponentType<P>,
  dsn: string | undefined = process.env.EXPO_PUBLIC_SENTRY_DSN,
): ComponentType<P> {
  return dsn ? (Sentry.wrap(Root) as ComponentType<P>) : Root;
}

File: services/api/app/observability/crash_reporting.py
"""Crash reporting (Sentry) — on only when `SENTRY_DSN` is set, and never with PII.

02 §8's rule for logs holds for crash reports, which are log lines that leave
the building: `user_id` and `request_id`, **never** emails, bodies, tokens or
image contents. Sentry's own switches do most of it (`send_default_pii=False`,
no request bodies, no frame locals); `scrub_event` is the part that does not
depend on remembering which switch covers which field. It runs on every event
and returns a **new** event — what it was handed is left untouched.

Off by default, like the AI provider (D23): `git clone` → `pytest` sends nothing
anywhere. The API and the worker initialise separately (D25) and tag themselves,
so a report says which process crashed.
"""
from __future__ import annotations

import re
from typing import Any

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration

from app.config import Settings

#: Headers worth keeping. An allow-list, because a deny-list misses the header
#: nobody thought of — `X-Forwarded-For` is the client's IP address.
_KEPT_HEADERS = frozenset({
    "accept", "content-length", "content-type", "host", "user-agent", "x-request-id",
})

#: Request fields that are the user's data or their address.
_DROPPED_REQUEST_FIELDS = frozenset({"data", "cookies", "env", "query_string"})

#: Keys whose VALUE is a secret, wherever they appear.
_SECRET_KEY = re.compile(r"password|passwd|secret|token|authorization|cookie|api_?key|signature",
                         re.IGNORECASE)

#: Keys dropped outright: frame locals, and a breadcrumb's query and fragment.
_DROPPED_KEYS = frozenset({"vars", "http.query", "http.fragment"})

_EMAIL = re.compile(r"[\w.+-]+@[\w-]+(?:\.[\w-]+)+")
_JWT = re.compile(r"eyJ[\w-]+\.[\w-]+\.[\w-]+")
_URL_QUERY = re.compile(r"(https?://[^\s?#\"']+)[?#][^\s\"']*")

FILTERED = "[Filtered]"


def scrub_event(event: dict[str, Any], hint: dict[str, Any] | None = None) -> dict[str, Any]:
    """`before_send`: the event without personal data. Never drops it."""
    cleaned = _scrub(event)
    request = cleaned.get("request")
    if isinstance(request, dict):
        cleaned = {**cleaned, "request": _scrub_request(request)}
    user = cleaned.get("user")
    if isinstance(user, dict):
        # The id is how a report is joined to a support request; nothing else.
        cleaned = {**cleaned, "user": {k: v for k, v in user.items() if k == "id"}}
    return cleaned


def _scrub_request(request: dict[str, Any]) -> dict[str, Any]:
    kept = {k: v for k, v in request.items() if k not in _DROPPED_REQUEST_FIELDS}
    headers = kept.get("headers")
    if isinstance(headers, dict):
        kept = {**kept, "headers": {
            k: v for k, v in headers.items() if k.lower() in _KEPT_HEADERS
        }}
    return kept


def _scrub(value: Any) -> Any:
    if isinstance(value, dict):
        return {
            k: (FILTERED if _SECRET_KEY.search(str(k)) else _scrub(v))
            for k, v in value.items()
            if k not in _DROPPED_KEYS
        }
    if isinstance(value, list | tuple):
        return [_scrub(v) for v in value]
    if isinstance(value, str):
        return _scrub_text(value)
    return value


#: SQLAlchemy's "[parameters: (...)]" suffix on a database error. The engines
#: hide it already; this is the second layer for anything they do not build.
_SQL_PARAMETERS = re.compile(r"\[parameters: .*?\]", re.DOTALL)


def _scrub_text(text: str) -> str:
    text = _SQL_PARAMETERS.sub("[parameters: redacted]", text)
    # The query first: an upload URL's query string IS its signature.
    text = _URL_QUERY.sub(r"\1", text)
    text = _JWT.sub("[token]", text)
    return _EMAIL.sub("[email]", text)


def init_crash_reporting(settings: Settings, *, process: str, transport: Any = None) -> bool:
    """Initialises Sentry for `process` ("api" or "worker"). Returns whether it did.

    `transport` is for the suite, which captures events instead of sending them.
    """
    if not settings.sentry_dsn:
        return False
    sentry_sdk.init(
        dsn=settings.sentry_dsn,
        environment=settings.environment,
        release=settings.app_release or None,
        send_default_pii=False,
        include_local_variables=False,
        max_request_body_size="never",
        before_send=scrub_event,
        before_send_transaction=scrub_event,
        # Errors only. Tracing is a second data stream with its own PII surface,
        # and 02 §9's RED metrics already answer "how slow".
        traces_sample_rate=0.0,
        integrations=[StarletteIntegration(), FastApiIntegration()],
        transport=transport,
    )
    # The global scope, so every event carries it — each request's own scope
    # is forked later, from a context this call may not be in.
    sentry_sdk.get_global_scope().set_tag("process", process)
    return True


def report_unhandled(exc: BaseException, *, request_id: str) -> None:
    """One event per unhandled error, tagged with the id the user was shown —
    the thread from a support email to the stack trace (docs/06 §10). A no-op
    when crash reporting is off."""
    with sentry_sdk.new_scope() as scope:
        scope.set_tag("request_id", request_id)
        scope.capture_exception(exc)
